app/assets/javascripts/libs/diffable_map.js

```diff
@@ -0,0 +1,294 @@
+// @flow
+// import Utils from "libs/utils";
+
+const defaultItemsPerBatch = 10000;
+let idCounter = 0;
+
+// Some properties:
+// Insertion order is not guaranteed
+
+class DiffableMap<K: number, V> {
+  id: number;
+  chunks: Array<Map<K, V>>;
+  entryCount: number;
+  existsCache: Map<K, boolean>;
+  itemsPerBatch: number;
+
+  constructor(optKeyValueArray?: ?Array<[K, V]>, itemsPerBatch: number = defaultItemsPerBatch) {
+    // Make the id property not enumerable so that it does not interfere with tests
+    Object.defineProperty(this, "id", { value: idCounter++, writable: true });
+    // this.id = idCounter++;
+    this.chunks = [];
+    this.existsCache = new Map();
+    this.entryCount = 0;
+    this.itemsPerBatch = itemsPerBatch;
+
+    if (optKeyValueArray != null) {
+      for (const [key, value] of optKeyValueArray) {
+        this.mutableSet(key, value);
+      }
+    }
+  }
+
+  get(key: K): V {
+    let idx = 0;
+    while (this.chunks[idx] != null) {
+      if (this.chunks[idx].has(key)) {
+        // $FlowFixMe
+        return this.chunks[idx].get(key);
+      }
+      idx++;
+    }
+    throw new Error("Get empty");
+  }
+
+  has(key: K): boolean {
+    try {
+      this.get(key);
+      return true;
+    } catch (exception) {
+      return false;
+    }
+  }
+
+  set(key: K, value: V): DiffableMap<K, V> {
+    if (this.existsCache.has(key)) {
+      let idx = 0;
+      while (this.chunks[idx] != null) {
+        if (this.chunks[idx].has(key)) {
+          const newMap = shallowCopy(this);
+          newMap.chunks[idx] = new Map(this.chunks[idx]);
+          newMap.chunks[idx].set(key, value);
+          return newMap;
+        }
+        idx++;
+      }
+      throw new Error("should not happen");
```

Comment:
"What" should not happen ;-)

Reply:
The code path should never be reached due to the control flow. The error message is to ensure Flow that the annotated return type is correct. I'll update the message to reflect this better.